Add EIP-4910: Royalty Bearing NFTs (#4910)
ethereum/EIPs: EIPS/eip-xxxx.md

---

> EIP: TBD

Reviewer: ```suggestion
eip: 4910
```

Author: updated

> Title: Royalty Bearing NFTs

Reviewer: ```suggestion
title: Royalty Bearing NFTs
```

Author: done

> Description: This proposal outlines the structure of an extension of the ERC721 standard for NFTs to correctly define, process, and pay (hierarchical) onchain royalties from NFT sales, and goes beyond [EIP-2981](https://eips.ethereum.org/EIPS/eip-2981).

Reviewer: ```suggestion
description: This proposal outlines the structure of an extension of the ERC721 standard for NFTs to correctly define, process, and pay (hierarchical) onchain royalties from NFT sales, and goes beyond [EIP-2981](https://eips.ethereum.org/EIPS/eip-2981).
```

Reviewer: This is too long, and shouldn't link to anything.  `ERC721` should be `EIP-721` as well.

> Author: Andreas Freund (@Therecanbeonlyone1969) 

Reviewer: ```suggestion
author: Andreas Freund (@Therecanbeonlyone1969) 
```

discussions-to: https://ethereum-magicians.org/t/royalty-bearing-nfts/8453

Status: Draft

Type: Standards Track

Category: ERC

Created: 2022-03-14

> Requires: ERC721, ERC165, ERC2771

Reviewer: No newlines between each of these.

Reviewer: Also the field names should be all lower-case.

Author: done

> Requires: ERC721, ERC165, ERC2771

Reviewer: ```suggestion
requires: 165, 721, 2771
```

Author: done

The proposal builds upon the [OpenZeppelin Smart Contract Toolbox](https://github.com/OpenZeppelin/openzeppelin-contracts) architecture, and extends it to include royalty account management (CRUD), royalty balance and payments management, simple trading capabilities -- Listing/De-Listing/Buying -- and capabilities to trace trading on exchanges. The royalty management capabilities allow for hierarchical royalty structures, referred to herein as royalty trees, to be established by logically connecting a "parent" NFT to its "children", and recursively enabling NFT "children" to have more children. 

## Motivation
The management of royalties is an age-old problem characterized by complex contracts, opaque management, plenty of cheating and fraud. 

The above is especially true for a hierarchy of royalties, where one or more assets is derived from an original asset such as a print from an original painting, or a song is used in the creation of another song, or distribution rights and compensation are managed through a series of affiliates. 

In the example below, the artist who created the original is eligible to receive proceeds from every sale, and resale, of a print. 

> ![Fig1](https://i.imgur.com/Py6bYQw.png)

Reviewer: Figures should be added to `../assets/eip-4910/*` and be linked with a relative link.

The basic concept for hierarchical royalties utilizing the above "ancestry concept" is demonstrated in the figure below.

> ![Fig2](https://i.imgur.com/7MtWzBV.png)

Reviewer: Same as above, don't externally link to images.

        - Following the same approach as above but now targeting the `addListNFT` function, an attacker can force an out of gas error or run time errors in the `executePayment` function by listing many NFTs at a low price, and then performing a purchase from another account. Mitigation: Limit the number of NFTs that can be included in one listing.
        - The creator of the NFT family could set the number of generations too high such that the royalty distribution function could incur and out of gas or run time error because of the recursive nature of the function. Mitigation: Limiting the `maxNumberGeneration` by the creator.
    - General Considerations: The creator of an NFT family must carefully consider the business model for the NFT family and then set the parameters such as maximum number of generations, royalty sub-accounts, number of prints per print, number of NFTs in a listing, and the maximum and minimum royalty split percentage allowed. 

- **Phishing Attacks**

>     - NFT phishing attacks often target the `approve` and `setApprovalForAll` functions by tricking owners of NFTs to sign transactions adding the attacker account as approved for one or all NFTs of the victim. Mitigation: This contract is not vulnerable to these type of phishing attacks because all NFT transfers are sales, and the NFT contract itself is the owner of all NFTs. This means that transfers after a purchase are achieved by setting the new owner in the `_approve` function. Calling the public 'approve' function will cause the function call to error out because `msg.sender' of the malicious transaction cannot be the NFT owner.

Reviewer: ```suggestion
    - NFT phishing attacks often target the `approve` and `setApprovalForAll` functions by tricking owners of NFTs to sign transactions adding the attacker account as approved for one or all NFTs of the victim. Mitigation: This contract is not vulnerable to these type of phishing attacks because all NFT transfers are sales, and the NFT contract itself is the owner of all NFTs. This means that transfers after a purchase are achieved by setting the new owner in the `_approve` function. Calling the public `approve` function will cause the function call to error out because `msg.sender` of the malicious transaction cannot be the NFT owner.
```

Author: corrected
